### ISSUE TITLE
error openDatabase cannot open file.

### ISSUE THREAD

**Issue by tedhenry100** [tekartik/sqflite#191]
On a physical Samsung Galaxy SIII running Android 4.4.2, I cannot open a database. I know it is an old device with an old Android but should it still work?

Things do seem to work on the Android simulator with Android 9.

My helper function is currently the following. (I'm just trying to do a proof of concept so it isn't polished.)

    Future<sqflite.Database> openDatabase() async {
      var databasesPath = await sqflite.getDatabasesPath();
      var path = databasesPath + '/db.sqlite3';
      var exists = await sqflite.databaseExists(path);
      if (!exists) {
        print('copying assets database 100');
        ByteData data =
            await rootBundle.load('assets/db.sqlite3');
        print('copying assets database 200');
        List<int> bytes =
            data.buffer.asUint8List(data.offsetInBytes, data.lengthInBytes);
        print('copying assets database 300');
        await File(path).writeAsBytes(bytes);
        print('copying assets database 400');
      }
      return await sqflite.openDatabase(path);
    }

I see the following in Android Studio's Logcat. Note that "copying assets database 400" doesn't appear in the log.

    04-14 12:00:37.836 3274-3427/com.company.example I/flutter: copying assets database 100
    04-14 12:00:38.436 3274-3427/com.company.example I/flutter: copying assets database 200
    04-14 12:00:38.436 3274-3427/com.company.example I/flutter: copying assets database 300
    04-14 12:00:38.746 3274-3427/com.company.example E/flutter: [ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: FileSystemException: Cannot open file, path = '/data/data/com.company.example/databases/db.sqlite3' (OS Error: No such file or directory, errno = 2)
        #0      _File.open.<anonymous closure> (dart:io/file_impl.dart:364:9)
        #1      _rootRunUnary (dart:async/zone.dart:1132:38)
        #2      _CustomZone.runUnary (dart:async/zone.dart:1029:19)
        #3      _FutureListener.handleValue (dart:async/future_impl.dart:126:18)
        #4      Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:639:45)
        #5      Future._propagateToListeners (dart:async/future_impl.dart:668:32)
        #6      Future._completeWithValue (dart:async/future_impl.dart:483:5)
        #7      Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:513:7)
        #8      _rootRun (dart:async/zone.dart:1124:13)
        #9      _CustomZone.run (dart:async/zone.dart:1021:19)
        #10     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
        #11     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
        #12     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
        #13     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)

When I look in the Android Studio's Device File Explorer I do see `data/data/com.company.example/database/db.sqlite3`.  If I save it to my filesystem, it even looks like the file was created correctly.

If I rebuild and install the app again on the Galaxy SIII, the Flutter app tries to copy the database again even though the file is there. On the simulator with Android 9, it does not try to copy the database again because the app correctly recognizes that the file already exists.

**Comment by alextekartik** [comment]
What you could try:
- try to create the parent directory before writing the file
- try to flush (write has a flush param) before opening the database file

**Comment by tedhenry100** [comment]
Wow! Creating the parent directory worked. Thank you! I'm curious. What gave you the idea that might solve the problem?

Adding only `flush: true` did not fix the problem but I've added it below anyway because it seems like quite a good idea. I do not know what might happen if the database is open but the write has not yet flushed. Thanks for this suggestion too.

My unpolished but working helper function is now

    Future<sqflite.Database> openDatabase() async {
      var databasesPath = await sqflite.getDatabasesPath();
      var path = databasesPath + '/db.sqlite3';
      var exists = await sqflite.databaseExists(path);
      if (!exists) {
        ByteData data = await rootBundle.load('assets/db.sqlite3');
        List<int> bytes = data.buffer.asUint8List(data.offsetInBytes, data.lengthInBytes);
        await Directory(databasesPath).create(recursive: true);
        await File(path).writeAsBytes(bytes, flush: true);
      }
      return await sqflite.openDatabase(path);
    }

**Comment by alextekartik** [comment]
> What gave you the idea that might solve the problem?

That is typically the main reason why writing a file failed (and this is not related to sqflite, it is a just an io issue), and also others had similar issues where the databases folder was no created in some edge case scenario. Your "unpolished" function looks better now!

**Comment by kaiserleka** [comment]
I have tested on S3 simuators (api 16,17,18,19). My app with sqflite have worked only on api 19 and above.

So I have changed "minSdkVersion" from 16 to 19

**Comment by k00na** [comment]
@kaiserleka i think that Flutter in general can't be ran on simulators with Android API versions of 16, 17, 18... these versions of emulators don't have the ability to run the Flutter engine

**Comment by baeckerg** [comment]
I seem to have a similar problem on a MacOS ARM machine. My very simple code (I admit I am a newbie) looks like this

`final db = '/Users/gernot/Documents/Delphi/test.db3';

void openDB() async {
  var exists = await databaseExists(db);
  //var databasesPath = await getDatabasesPath();
  if (!exists) {
    print('Does not exist');
  } else {
    print('Exists!!!');
    var localDB = await openDatabase(db);
  }
}`

and ends up with

2022-12-05 16:54:24.815 flutter_application_1[17316:503225] error opening!: 23
2022-12-05 16:54:24.815 flutter_application_1[17316:503225] Could not create database queue for path /Users/gernot/Documents/Delphi/test.db3
2022-12-05 16:54:24.815 flutter_application_1[17316:503225] Could not open db.

Any help would be appreciated

**Comment by V3ntus** [comment]
Building with Mac OS, I could not find any other reports on this, until I narrowed it down to Mac OS application entitlements. Flutter, by default, sets the app sandbox key to true which restricts file access to user selected. It made sense when I could only open the file after I selected it again in my app's file open dialog.

Setting this key to `false` fixes it, but be aware of the repercussions when building your Release:
`/macos/Runner/DebugProfile.entitlements`
```xml
<key>com.apple.security.app-sandbox</key>
<false/>
```


<sub>I wish I saw #917 sooner. I spent many days banging my head on this</sub>
